base/rational.jl

```diff
@@ -10,23 +10,61 @@ struct Rational{T<:Integer} <: Real
     num::T
     den::T
 
-    function Rational{T}(num::Integer, den::Integer) where T<:Integer
-        num == den == zero(T) && __throw_rational_argerror_zero(T)
-        num2, den2 = divgcd(num, den)
-        if T<:Signed && signbit(den2)
-            den2 = -den2
-            signbit(den2) && __throw_rational_argerror_typemin(T)
-            num2 = -num2
+    function Rational{T}(num::Integer, den::Integer; checked::Bool) where T<:Integer
+        if checked
+            num == den == zero(T) && __throw_rational_argerror_zero(T)
+            num, den = divgcd(num, den)
+            num, den = checking_den(num, den)
         end
-        return new(num2, den2)
+        return new{T}(num, den)
     end
 end
 @noinline __throw_rational_argerror_zero(T) = throw(ArgumentError("invalid rational: zero($T)//zero($T)"))
 @noinline __throw_rational_argerror_typemin(T) = throw(ArgumentError("invalid rational: denominator can't be typemin($T)"))
 
-Rational(n::T, d::T) where {T<:Integer} = Rational{T}(n,d)
-Rational(n::Integer, d::Integer) = Rational(promote(n,d)...)
-Rational(n::Integer) = Rational(n,one(n))
+function checking_den(num::T, den::T) where T<:Integer
```

Comment:
I can see this unwanted behavior, which can be easily improved:
```
julia> Rational{Int}(UInt(0), 1)
ERROR: MethodError: no method matching checking_den(::UInt64, ::Int64)
Closest candidates are:
  checking_den(::T, ::T) where T<:Integer at /home/julia/julia/base/rational.jl:25
Stacktrace:
 [1] Rational{Int64}(::UInt64, ::Int64; checked::Bool) at /home/julia/julia/base/rational.jl:18
```
by
```
function checking_den(num::Integer, den::Integer)
```

Reply:
The intention for this `::T` specification was to avoid erroring on stuff like
```julia
Rational(UInt(1), typemin(Int32))
```
by ensuring that the arguments have already been promoted. I thought promotion happened in `divgcd` but that's not always the case.
Actually the example right here is a bit tricky, I'm working on it.

Reply:
I think working with #35507 might actually help solve the issue so I'll rebase and check

Reply:
OK so no, that won't work unless I relax even more the type constraints of `gcd` like so:
```julia
function gcd(a::U, b::S) where {U<:Unsigned, S<:Signed}
    T = promote_type(signed(U), S)
    gcd(promote(a, checked_abs(T(b)))...)
end
function gcd(a::S, b::U) where {U<:Unsigned, S<:Signed}
    T = promote_type(signed(U), S)
    gcd(promote(checked_abs(T(a)), b)...)
end
```
instead of the current
```julia
gcd(a::Unsigned, b::Signed) = gcd(promote(a, abs(b))...)
gcd(a::Signed, b::Unsigned) = gcd(promote(abs(a), b)...)
```

That's beyond the point of this PR so I'm going to use your version for the time being, and maybe I'll PR this change afterwards.

Reply:
To react to @JeffreySarnoff's comment just below, I think the minimal fix is simply to replace
```julia
num, den = checking_den(num, den)
```
in the inner constructor by
```julia
num, den = checking_den(T(num), T(den))
```

Reply:
Fixed in the latest commit, along with tests.
The specific case of `Rational(UInt(1), typemin(Int32))` is now a `@test_broken` (compared to being broken with no test).